torch/csrc/lazy/core/ir.h

```diff
@@ -161,19 +178,42 @@ class TORCH_API Node {
   // The IR framework user can attach a user defined metadata object deriving
   // from UserMetaData.
   std::shared_ptr<UserMetaData> user_metadata_;
-};
-
 
+  size_t node_list_index_;
+};
 
 inline std::ostream& operator<<(std::ostream& stream, const Node& node) {
   stream << node.ToString();
   return stream;
 }
 
+using OpList = c10::ArrayRef<Value>;
+
+class Std;
+
+template <typename T, typename... Args>
+NodePtr LookupNode(Args... args) {
+  size_t currrent_index = Node::node_list.size();
+  if (currrent_index >= Node::last_node_list.size()) {
+    return nullptr;
+  }
+  NodePtr node = Node::last_node_list[currrent_index];
+  T* ptr = dynamic_cast<T*>(node.get());
+  return (ptr && ptr->Equal(std::forward<Args>(args)...)) ? node : nullptr;
+}
+
 // TODO(alanwaketan): Support r-value reference argument type.
 template <typename T, typename... Args>
 NodePtr MakeNode(Args&&... args) {
-  return std::make_shared<T>(std::forward<Args>(args)...);
+  NodePtr node = LookupNode<T>(args...);
+  if (!node) {
+    TORCH_LAZY_COUNTER("IrNodeCreated",1);
+    node = std::make_shared<T>(std::forward<Args>(args)...);
+  } else {
+    TORCH_LAZY_COUNTER("IrNodeReused",1);
+  }
+  Node::PushIntoNodeList(node);
```

Comment:
Are you sure if all the places that you pass a node around are avoiding to invoke the node's copy constructor?